renn/data/tokenizers.py

```diff
@@ -15,6 +18,114 @@
 UNK = '<unk>'
 CLS = '<cls>'
 SEP = '<sep>'
+EOS = '<eos>'
+BOS = '<bos>'
+
+
+def _get_transform_fn(lower: bool, separate_punctuation: bool):
+  """Helper function for text_generator"""
+
+  if not lower and not separate_punctuation:
+    return lambda x: x
+  elif lower and not separate_punctuation:
+    return lambda x: x.lower()
+  elif not lower and separate_punctuation:
+    return lambda x: _punctuation_separator(x)
+  elif lower and separate_punctuation:
+    return lambda x: _punctuation_separator(x.lower())
+
+
+def _punctuation_separator(s: str) -> str:
+  """Separates punctuation at the end of word and end of line"""
+  punctuation_chars = ['.', ',', ':', ';', '?', '!']
+  special_chars = ['.', '?']  # regex chars
+  s_ = s
+  for c in punctuation_chars:
+    if c in special_chars:
+      # separate punctuation at end-of-line
+      s_ = re.sub(f'\{c}$', f' {c}', s_)
+      # separate punctuation at end-of-word
+      s_ = re.sub(f'\{c} ', f' {c} ', s_)
+    else:
+      # separate punctuation at end-of-line
+      s_ = re.sub(f'{c}$', f' {c}', s_)
+      # separate punctuation at end-of-word
+      s_ = re.sub(f'{c} ', f' {c} ', s_)
+  return s_
+
+
+def _tensor_punctuation_separator(x: tf.Tensor) -> tf.Tensor:
+  """Separates punctuation at the end of word and end of line.
```

Comment:
nit: Separate the first line of the docstring from the rest with an empty line.

e.g.:

```
def foo():
  """The first line description should be separate from the rest.

  More details about the function. This next section
  can span multiple lines, if needed.
  """
```

(the reason for being nitpick-y here is that the documentation [website](https://reverse-engineering-neural-networks.readthedocs.io) scrapes docstrings to generate the website, which need to be in a particular format)